### PR TITLE
add ruby String to rust Byte Collection conversion

### PR DESCRIPTION
**Pull request by cramt**
i Added support for converting a ruby String into a rust byte collection

In ruby the main way to represent some bytes is with a string, (see File.read or Base64.strict_decode64). and the current way to accept such
a string with magnus is by receiving a RString and calling `.as\_slice().as\_vec()` which is unsafe, this creates safe interface to receive bytes from ruby.

I've decided to use the Bytes crate since it seems to be rust communty's prefered way of dealing with collections of bytes and made the inclusion of it a non defaulted feature

### CODE REVIEW

File: src/r_string.rs [hunk 1]
```diff
@@ -900,6 +900,24 @@ impl RString {
...
+    /// ```
+    /// use magnus::RString;
+    /// use bytes::Bytes;
+    /// # let _cleanup = unsafe { magnus::embed::init() };
+    ///
+    /// let s = RString::new("example");
+    /// assert_eq!(s.to_bytes(), Bytes::from("example"));
+    /// ```
+    #[cfg(feature = "bytes-crate")]
+    pub fn to_bytes(self) -> bytes::Bytes {
```
Comment (matsadler) [inline src/r_string.rs:916]: If you add `#[cfg_attr(docsrs, doc(cfg(feature = "bytes-crate")))]` here the docs on docs.rs will call out that this is only available with the `bytes-crate` feature.

If you want to see what this looks like locally you can approximate the docs.rs build with:
``` sh
env RUSTDOCFLAGS='--cfg docsrs' cargo +nightly doc --all-features
```

File: Cargo.toml [hunk 2]
```diff
@@ -21,10 +21,12 @@ exclude = ["examples/rust_blank/ext/rust_blank", "examples/custom_exception_ruby
 embed = ["rb-sys/link-ruby"]
 rb-sys-interop = []
 ruby-static = ["rb-sys/ruby-static"]
+bytes-crate = ["bytes"]
 
 [dependencies]
 magnus-macros = { version = "0.4.0", path = "magnus-macros" }
 rb-sys = { version = "0.9.56", default-features = false, features = ["bindgen-rbimpls", "bindgen-deprecated-types"] }
+bytes = { version = "1.4.0", optional = true }
```
Comment (matsadler) [inline Cargo.toml:29]: Could you use `version = "1"` for the version here? That way if someone is using a crate with Magnus that has a dependency on an older (but still 1.x.x) version of `bytes` then cargo can just use that single version across the two crates.

**Review by matsadler — COMMENTED** [review]
Thanks, this is great!

Could you add an `IntoValue` implementation for `Bytes` too? That way it can also be returned to Ruby.
I'm in the process of transitioning from `Into<Value>` to `IntoValue`, but `Into<Value>` won't go away til after the next release, so there should probably be a `impl From<bytes::Bytes> for Value` too.

I'll add an entry to the CHANGELOG.md for this after I merge, but feel free to add it there yourself in your own words if you want to.

File: src/try_convert.rs [hunk 3]
```diff
@@ -167,6 +167,17 @@ impl TryConvert for String {
 }
 impl TryConvertOwned for String {}
 
+#[cfg(feature = "bytes-crate")]
+impl TryConvert for bytes::Bytes {
+    fn try_convert(val: Value) -> Result<bytes::Bytes, Error> {
+        debug_assert_value!(val);
+        Ok(RString::try_convert(val)?.to_bytes())
```
[the remark this answers is not in the document]
Comment (cramt) [reply src/try_convert.rs:174]: well, if we add that restriction we might wanna also add a UTF-8 encoding restriction when converting a RString into String, though this is of course a breaking change

[the remark this answers is not in the document]
Comment (matsadler) [reply src/try_convert.rs:174]: So there's a pretty strong precedent for calling `#to_str` when a `String` is desired, as that's what Ruby does internally, and what the `StringValue` macro does for C extensions. And I went with re-encoding to UTF-8 automatically as I think the whole `default_internal`/`default_external` mechanism is a good precedent for automatic re-encoding being the way Ruby is expected to work.

My initial thought was there isn't really a direct equivalent to `Bytes` in C, so there isn't an obvious "this is the way Ruby would do it" path to follow.

I can't think of that many Ruby APIs that expect binary strings. The best I could come up with were `Base64::encode64`, `Array#pack` and a bunch of stuff in the OpenSSL bindings. All these seem to use `StringValue` and ignore the encoding, so I guess what you've got here matches that.

**Comment by matsadler** [comment]
Thank you!

**Pull request merged**